### PR TITLE
C#: add store step for return statements inside async methods

### PR DESCRIPTION
**Pull request by tamasvajk**
Fixes https://github.com/github/codeql/issues/5410.

~~[Differences job](https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/944/)~~ [Differences job](https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/948/)

### CODE REVIEW

File: csharp/ql/src/semmle/code/csharp/Callable.qll [hunk 1]
```diff
@@ -206,7 +207,11 @@ class Callable extends DotNet::Callable, Parameterizable, ExprOrStmtParent, @cal
     exists(ReturnStmt ret | ret.getEnclosingCallable() = this | e = ret.getExpr())
     or
     e = this.getExpressionBody() and
-    not this.getReturnType() instanceof VoidType
+    not this.getReturnType() instanceof VoidType and
+    (
+      not this.(Modifiable).isAsync() or
+      not this.getReturnType() instanceof SystemThreadingTasksTaskClass
```
Comment (tamasvajk) [inline csharp/ql/src/semmle/code/csharp/Callable.qll:213]: This is not entirely correct. After C#7, it is enough to have a return type that has a `GetAwaiter` method.

File: csharp/ql/src/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll [hunk 2]
```diff
@@ -1397,6 +1406,30 @@ private module ReturnNodes {
     override string toStringImpl() { result = yrs.toString() }
   }
 
+  /**
+   * A synthesized `return` node for returned expressions inside `async` methods.
+   */
+  class AsyncReturnNode extends ReturnNode, NodeImpl, TAsyncReturnNode {
```
Comment (tamasvajk) [inline csharp/ql/src/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:1412]: At the moment we are not handling async streams. See https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/async/async-return-types#async-streams-with-iasyncenumerablet

File: csharp/ql/src/semmle/code/csharp/Callable.qll (hunk 1 above)
Comment (tamasvajk) [reply csharp/ql/src/semmle/code/csharp/Callable.qll:213]: It is probably enough to require a generic type.

**Review by hvitved — APPROVED** [review]
Excellent, Tamas!

**Pull request merged**